### PR TITLE
RSDK-7721: Fix bad module startup hang

### PR DESCRIPTION
**Pull request by maximpertsov**
The RDK server hangs and cannot process requests while it tries to load a module with a fatal bug (e.g. syntax error, immediate panic). This PR tries to detect if a module cannot start due to such an error and aborts loading to unblock the RDK.

### Testing
* See https://github.com/viamrobotics/rdk/pull/4090#issuecomment-2180710730

### CODE REVIEW

File: module/modmanager/manager.go [hunk 1]
```diff
@@ -1106,6 +1106,9 @@ func (m *module) startProcess(
 			}
 			return ctxTimeout.Err()
 		case <-checkTicker.C:
+			if errors.Is(m.process.Status(), os.ErrProcessDone) {
+				return fmt.Errorf("module %s cannot start", m.cfg.Name)
```
Comment (maximpertsov) [inline module/modmanager/manager.go:1110]: this error message should give more guidance on what's wrong/what to do

Comment (benjirewis) [reply module/modmanager/manager.go:1110]: Maybe "module %s exited quickly after attempted startup; your module may have run into an immediate runtime issue" or something?

File: module/modmanager/manager_test.go [hunk 2]
```diff
@@ -1317,3 +1317,25 @@ func TestAddStreamMaxTrackErr(t *testing.T) {
 	test.That(t, err.Error(), test.ShouldContainSubstring, "only 9 WebRTC tracks are supported per peer connection")
 	test.That(t, sub, test.ShouldResemble, rtppassthrough.NilSubscription)
 }
+
+func TestBadModuleFailsFast(t *testing.T) {
+	t.Setenv("TESTMODULE_PANIC", "1")
+	logger, _ := logging.NewObservedTestLogger(t)
```
Comment (benjirewis) [inline module/modmanager/manager_test.go:1323]: [nit] Might as well just use `NewTestLogger`.

**Review by benjirewis — COMMENTED** [review]
Nice!

File: module/modmanager/manager.go [hunk 3]
```diff
@@ -1106,6 +1106,9 @@ func (m *module) startProcess(
 			}
 			return ctxTimeout.Err()
 		case <-checkTicker.C:
+			if errors.Is(m.process.Status(), os.ErrProcessDone) {
```
[the remark this answers is not in the document]
Comment (maximpertsov) [reply module/modmanager/manager.go:1109]: I'm going to do a bit more testing here, but so far that's the only error I've gotten in cases where a module panics/fails irrecoverably.

File: module/modmanager/manager_test.go (hunk 2 above)
Comment (maximpertsov) [reply module/modmanager/manager_test.go:1323]: ah yes, definitely

File: module/modmanager/manager.go (hunk 1 above)
Comment (maximpertsov) [reply module/modmanager/manager.go:1110]: ideally we could get some more information from the process crash... but beyond that your suggestion is an improvement - I essentially want to encourage the user to inspect their module definition for any "obvious" reasons for an immediate fatal crash

**Comment by maximpertsov** [comment]
Tested on a real robot running viam and was able to demonstrate the fix - I added a module that required a shared library to be linked at runtime:

```
Jun 20 13:26:28 REDACTED viam-server[698]: \_ REDACTED/bin/testmodule: REDACTED/libc.so.6: version `GLIBC_2.32' not found (required by REDACTED/bin/testmodule)
```

Running this module with the current version of RDK results in a hang with the message `Waiting for module to complete startup and registration`. However, the patched version of RDK was able to start without waiting for the broken module to start.

File: module/modmanager/manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximpertsov) [reply module/modmanager/manager.go:1109]: I [tested on a real robot and a different kind of module failure](https://github.com/viamrobotics/rdk/pull/4090#issuecomment-2180710730) and still only see this error. I'm a bit more convinced that keying on this error is sufficient.

File: module/modmanager/manager.go (hunk 1 above)
Comment (maximpertsov) [reply module/modmanager/manager.go]: went with a version of your suggestion for now https://github.com/viamrobotics/rdk/pull/4090/commits/bec0fc347daae389221dbbd8c026d86be6ee18f2

the fatal runtime error should also be logged on a nearby line when this happens so hopefully this is an improvement overall

File: module/modmanager/manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply module/modmanager/manager.go:1109]: Nice thanks for further testing; seems fine to me.

File: module/testmodule/main.go [hunk 4]
```diff
@@ -32,6 +32,9 @@ func main() {
 }
 
 func mainWithArgs(ctx context.Context, args []string, logger logging.Logger) error {
+	if os.Getenv("VIAM_TESTMODULE_PANIC") != "" {
+		panic("there is no flavor town")
```
Comment (benjirewis) [inline module/testmodule/main.go:36]: 😆 is this a reference to something?

**Review by benjirewis — APPROVED** [review]

File: module/modmanager/manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximpertsov) [reply module/modmanager/manager.go:1109]: So one kind of error that this patch does not address is scripts that do not crash, but cannot receive a socket address. See [my ticket comment](https://viam.atlassian.net/browse/RSDK-7721?focusedCommentId=31508) for more details. I think we can address this separately.

File: module/testmodule/main.go (hunk 4 above)
Comment (maximpertsov) [reply module/testmodule/main.go:36]: hah, forgot i used this... I imagine this is a valid reason to panic for some people

File: module/modmanager/manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply module/modmanager/manager.go:1109]: Yeah... I think it's gonna be difficult to distinguish between a module that just takes a while to start up and one that will never take ownership of its supplied sock file.

**Pull request merged**